pyroomacoustics/open_sofa_interpolate.py

```diff
@@ -0,0 +1,556 @@
+
+
+                no_of_nodes = file_sofa.Dimensions.M  # Number of measurement points
+
+                # samples = file_sofa.Dimensions.N  # Samples per IR
+                samples = rcv_msr.shape[1]
+
+                # All measurements should be in = radians phi [0,2*np.pi] , theta [0,np.pi]
+
+                phi_src = src_pos[:, 0]
+                theta_src = src_pos[:, 1]
+                r_src = src_pos[:, 2]
+
+                no_of_lat = list(collections.Counter(theta_src).values())[0]
+
+                #no_of_long = len(no_of_lat)
+
+                theta_16 = np.array([theta_src[i] for i in range(len(theta_src)) if i % no_of_lat == 0])
+
+                return phi_src, theta_src, r_src, theta_16, rcv_msr, no_of_nodes, samples, no_of_lat
+
+    def cal_index_knn(self,azimuth,colatitude):
+
+        """
+        Calculates nearest neighbour for all the image source for the source and the receiver
+        This method exist so that we don't have to query each image source , nearest neighbour for all the incoming angles (mic) and departure angles(sources)
+        are calculated at once.
+        : Parameter
+        ------------------------------
+
+        azimuth:nd.array() (int)  [-pi,pi]
+        colatitude:nd.array() (int) [0,pi]
+
+        : Return : nd.array() return nearest index on the grid based on the input angles using kDtree
+        """
+
+        longitude=azimuth.reshape(-1,1)
+        latitude=colatitude.reshape(-1,1)
+
+        start=timer()
+        if self.interpolate:
+            dd,ii=self.nn_kd_tree_rotated_fibo_grid.query(np.hstack((longitude,latitude)))
+        else:
+            dd, ii = self.nn_kd_tree_rotated_sofa_grid.query(np.hstack((longitude,latitude)))  # Query on the rotated set of points
+        end=timer()
+        print("Time taken For Query Once ", end - start)
+        return ii
+
+    def neareast_neighbour(self,index):
+        """
+        Interpolated or non interpolated response given the azimuth and colatitude of the angle of arrival (microphone) / angle of departure (source)
+        Query on a KDTree given azimuth and elevation and find the most neareast point according to that on the sphere.
+
+        :Parameter
+        -------------------------
+        azimuth: (int)  [-pi,pi]
+        colatitude: (int) [0,pi]
+
+        :Return (np.array) [Coeffs of DFT]
+        -------------------------
+        FIR in frequency resolution for the particular query angle.
+
+        """
+
+        if self.interpolate:
+            #dd, ii = self.nn_kd_tree_rotated_fibo_grid.query([longitude, latitude])  #Query on the rotated set of points
+            return self.sh_coeffs_expanded_target_grid[index, :]
+        else:
+            #dd, ii = self.nn_kd_tree_rotated_sofa_grid.query([longitude, latitude])  # Query on the rotated set of points
+            return self.sofa_msr_fir[index, :]
+
+    def change_orientation(self, azimuth_change, colatitude_change):
```

Comment:
I think this is why I suggested to do the rotation when computing the response for directions rather than on the grid itself.

Reply:
Actully, as I wrote above, we need to handle arbitrary rotations with three angles. Also, this code assumes the starting point is at [1, 0, 0]. If you apply it twice, the result might not be as expected.
That is
```python
obj2 = obj  # same starting point

obj.change_orientation(az1, col1)
obj.change_orientation(az2, col2)
# is different than
obj2.change_orientation(az1 + az2, col1 + col2)
```

Reply:
This is easy to see if you recall that matrices do not commute.